Networking guide (#1387)
jellyfin/jellyfin.org: docs/general/post-install/networking/8_reverse-proxy/index.md

:::caution

In order for a reverse proxy to have the maximum benefit, you should have a publicly routable IP address and a domain with DNS set up correctly.
These examples assume you want to run Jellyfin under a sub-domain (e.g. jellyfin.example.com), but are easily adapted for the root domain if desired.

:::

:::caution

Be careful when logging requests with your reverse proxy. Jellyfin sometimes sends authentication information as part of the URL (e.g `api_key` parameter), so logging the full request path can expose secrets to your logfile.
We recommend that you either protect your logfiles or do not log full request URLs or censor sensitive data from the logfile.
The nginx documentation below includes an example how to censor sensitive information from a logfile.

:::

Some popular options for reverse proxy systems are [Apache](https://httpd.apache.org), [Caddy](https://caddyserver.com), [Haproxy](https://www.haproxy.com), [Nginx](https://www.nginx.com) and [Traefik](https://traefik.io).

- [Apache](./apache)
- [Caddy](./caddy)
- [HAProxy](./haproxy)
- [Nginx](./nginx)

> - [Traefik](./traefik)

Reviewer: 🗿 most of the time people who know nothing is going to pick nginx and then fuck it up and then show up asking for troubleshooting help so please gently guide them towards caddy

Author: I still have not reworked the text of any of the reverse proxy guides. Hence this is still a draft 😅.

Noted though.

---
uid: reverse-proxy-index
title: Reverse Proxy
---

# Reverse Proxy

> ## General

Reviewer: this heading can be removed

Whilst a Proxy typically is meant to catch and forward outgoing traffic, a reverse proxy does the same for incoming network traffic.
The reverse Proxy can then act as an entrypoint to various services and will internally forward the traffic to your service.

This opens the ability to do rule specific routing, for example subdomain routing, ip geoblocking, ratelimits or url forwarding.
Its also possible to handle all DNS and SSL specific things on the reverse Proxy, so that everything routing wise can be managed in one place.

> Additionaly reverse proxies offer extended access logging, so its always clear who, when and where a network request came from and went to.

Reviewer: I don't think we really need to go this in depth on what a proxy and a reverse proxy is. Just 2-3 lines explaining what a reverse proxy does and why we are using it is enough

Author: The documentation page is not part of advanced networking and we recommend using a reverse proxy.
Therefore I absolutely see the need to go into detail when explaining reverse proxies!
A lot of users that will come over from Plex will need this kind of railing, as they are potentially new to doing networking by themselves.

Reviewer: I wouldn't expect normal users to know what a proxy is given they aren't really used much anymore

Reviewer: So at least cut out the parts mentioning a proxy server

## Running Jellyfin Behind a Reverse Proxy

Important things to note when using Jellyfin behind a reverse proxy.

### Logging

Be careful when logging requests with your reverse proxy. Jellyfin sometimes sends authentication information as part of the URL (e.g `api_key` parameter), so logging the full request path can expose secrets to your logfile.
We recommend that you either protect your logfiles or do not log full request URLs or censor sensitive data from the logfile.

> Our [nginx guide](./nginx/) includes an example how to censor sensitive information from a logfile.

Reviewer: 1. The only place this is a problem is going to be when the token is a parameter. Therefore only the URLs need to be discarded
2. The guide on how to censor urls in logs is going to be specific to each reverse proxy. Having everyone look at nginx docs is not good because they may very well be using something other than nginx

Author: Do the other guides also include examples on how to censor that info?
Also, you are putting it as if access tokens would rarely be send as the URL, however thats not the case!

### forwarded-for headers

Since the reverse proxy forwards the traffic to the Jellyfin-Server, Jellyfin will see all requests coming from the reverse proxy, not the actual Client IP's.
This introduces potential security risks and can also break compatibility.

Therefore the IP Address(es) of the reverse Proxy MUST be set in the Network configuration..
This will enable Jellyfin to search for `X-Forwarded-For`, `X-Forwarded-Proto` and `X-Forwarded-Host` headers and use the associated value as the source IP address.

This assumes that the reverse Proxy is set up to include this header, which is not always the case by default.

> If issues with source IP forwarding appear, this should be checked.

Reviewer: should also mention the known proxies setting in the dashboard and the consequences of not setting it up properly

### Websockets

> Jellyfin makes use of Websockets for various things. Not all reverse proxies allow this by default. Its important to make sure Websockets are allowed for your Jellyfin-Server.

Reviewer: ```suggestion
Jellyfin makes use of Websockets for various things. Not all reverse proxies allow this by default. Its important to make sure Websockets are allowed for your Jellyfin server.
```

> ## Guides

Reviewer 2: Nice work on this!! 

A reverse proxy guide is such an excellent addition to the networking guide

> We recommend using [Caddy](https://caddyserver.com/) for its easy install and use. We provide a [guide](./caddy/) for configuring Caddy with Jellyfin.

Reviewer: ```suggestion
We recommend using [Caddy](https://caddyserver.com/) for its ease of use, including automatic handling of TLS and websockets. We provide a [guide](./caddy/) for configuring Caddy with Jellyfin.
```

Author: What do you mean by "automatic handling of TLS and websockets"?
Websockets are allowed by default which is nice but not an outstanding caddy feature.

And "automatic TLS" still requires a lot of non-automatic setup, like certbot and the like.

Author: I still dissagree with you wording here. "automatic handling of websockets" - whats to handle about this? they only need to be ALLOWED, which is the default in most proxies.

And "automatic TLS" still requires you to have a domain and configure the reverse proxy accordingly, its not "magically working" just because you are using Caddy.

Reviewer: ```suggestion
We recommend using [Caddy](https://caddyserver.com/) for its ease of use, including automatic handling of TLS when a valid domain is provided. We provide a [guide](./caddy/) for configuring Caddy with Jellyfin.
```
I still think automatic https is a feature to highlight for the purpose of steering people towards caddy which is hard to fuck up instead of nginx

Reviewer: I think using the term https works here

If you do not want to use Caddy, some other popular options for reverse proxy systems are [Nginx](https://www.nginx.com), [Traefik](https://traefik.io), [Haproxy](https://www.haproxy.com) and [Apache](https://httpd.apache.org). Note that these have a greater learning curve than the recommended reverse proxy Caddy. You can find respective guides here:

- [Nginx](./nginx)
- [Traefik](./traefik)
- [HAProxy](./haproxy)
- [Apache](./apache)

While not a reverse proxy, Let's Encrypt can be used independently or with a reverse proxy to provide SSL certificates.

- [Let's Encrypt](../advanced/letsencrypt)

> 

Reviewer: if we mention lets encrypt here, fail2ban / ipban is also worth a mention

Author: I disagree, as SSL Certificates are a neccesity to set-up a reverse proxy.
Meanwhile fail2ban / ipban are nice to haves.
If you want them mentioned we could add them in the top where the advantages of a reverse proxy are explained

Reviewer: fair

When following these guides, be sure to replace the following variables with your information.

- `DOMAIN_NAME`: Your public domain name to access Jellyfin on (e.g. jellyfin.example.com)
- `example.com`: The domain name Jellyfin services will run under (e.g. example.com)
- `SERVER_IP_ADDRESS`: The IP address of your Jellyfin server (if the reverse proxy is on the same server use 127.0.0.1)

In addition, the examples are configured for use with Let's Encrypt certificates. If you have a certificate from another source, change the SSL configuration from `/etc/letsencrypt/DOMAIN_NAME/` to the location of your certificate and key.

> Ports 80 and 443 (pointing to the proxy server) need to be opened on your router and firewall.

Reviewer: this should be in the lets encrypt page IMO

Author: What??

Reviewer: I thought this is part of the letsencrypt section

---
uid: reverse-proxy-index
title: Reverse Proxy
---

## Reverse Proxy

> ### General

Reviewer: this heading doesn't need to exist

> ### General
> 
> A proxy server is meant to catch and forward outgoing traffic. A reverse proxy does the same for incoming network traffic.
> The reverse proxy can then act as an entrypoint to various services and will internally forward the traffic to your service.

Reviewer 3: ```suggestion
### General

A proxy server is meant to catch and forward outgoing traffic. A reverse proxy does the same, but for incoming network traffic.
It can serve as an entry point to various services and will internally forward the traffic to your service.
```

This opens the ability to do rule specific routing, for example subdomain routing, ip geoblocking, ratelimits or url forwarding.

> It is also possible to centralize DNS and SSL management on the reverse proxy, streamlining all routing-wise configuration.

Reviewer 3: TLS management.

Additionally, reverse proxies offer extended access logging, so it is always clear who, when and where a network request came from and went to.

## Running Jellyfin Behind a Reverse Proxy

Important things to note when using Jellyfin behind a reverse proxy.

### Logging

Be careful when logging requests with your reverse proxy. Jellyfin sometimes sends authentication information as part of the URL (e.g `api_key` parameter), so logging the full request path can expose secrets to your logfile.
We recommend that you either protect your logfiles or do not log full request URLs or censor sensitive data from the logfile.
Our [proxy guides](./#Guides) include examples on how to censor sensitive information from a logfile.

### Forwarded-For Headers

> When traffic is forwarded through a reverse proxy, Jellyfin sees the proxy’s IP rather than the client’s.

Reviewer: ```suggestion
When traffic is forwarded through a reverse proxy, Jellyfin sees the proxy’s IP address rather than the client’s IP address.
```

> This introduces potential security risks and can also break compatibility.

Reviewer 3: Is this "potential security risk" true? This text lets a user believe that jellyfin seeing a proxy ip is a security risk. Maybe explain more clearly what the security risk would be. I suppose this has something to do with "Known Proxies" ?

Reviewer 4: its a security risk because if you configured your RP to be in the local network zone and as all requests will then be seen as from the RP itself, you enable some features like server restart which yes is a security issue

Author: @felix920506 @JPVenson @TimGels 
Leave the wording like this, or do you want to suggest a change?
I can see Tims point that more context would be a good idea here.

Reviewer: I see it explained in the next paragraph but I think they should be the same paragraph. In addition, I think we should also explain why we only trust those headers from a source in the known proxies list

> This introduces potential security risks and can also break compatibility, since Jellyfin will not be abled to differenciate between local and remote access.

Author: Yes that. Also simply Jellyfins logging does not show when a malicious user has tried to log in. it also breaks compatibility with 90% of the networking features because they are just pointless if all requests are from the same IP.

> This introduces potential security risks and can also break compatibility, since Jellyfin will not be abled to differenciate between local and remote access.

Reviewer: Put that in the doc

> This introduces potential security risks and can also break compatibility, since Jellyfin will not be abled to differenciate between local and remote access.

Author: I have put "Therefore, if set up incorrectly, all limitations for external access will not work." now.
What do you say?

Therefore, the IP address(es) of your reverse proxy must be configured under “Known Proxies” in Jellyfin’s **Network** settings.

> This allows Jellyfin to respect the `X-Forwarded-For`, `X-Forwarded-Proto`, and `X-Forwarded-Host` headers and use the associated value as the source IP address. By default, Jellyfin will discard all forwarded-for headers if they come from an 'untrusted' source. This is so that malitious devices will not be abled to hide their IP by providing a forwarded-for header.

Author: We have trusted proxies.
Naturally if there is a proxy but it is NOT configured here, it's an 'untrusted' proxy

> This allows Jellyfin to respect the `X-Forwarded-For`, `X-Forwarded-Proto`, and `X-Forwarded-Host` headers and use the associated value as the source IP address. By default, Jellyfin will discard all forwarded-for headers if they come from an 'untrusted' source. This is so that malitious devices will not be abled to hide their IP by providing a forwarded-for header.

Author: By default, Jellyfin will discard all forwarded-for headers that do not originate from a "known Proxy".

This assumes that the reverse Proxy is set up to include this header, which is not always the case by default.
If issues with source IP forwarding appear, this should be checked.

### Websockets

Jellyfin makes use of Websockets for various things. Not all reverse proxies allow this by default. Its important to make sure Websockets are allowed for your Jellyfin-Server.

## Guides

> We recommend using [Caddy](https://caddyserver.com) for its easy install and use.

Reviewer: ```suggestion
We recommend using [Caddy](https://caddyserver.com) for most users.
```
maybe should link to the caddy page instead of the caddy project page

> You can find our Guide for using Jellyfin through Caddy [here](./caddy/).

Reviewer: you should not have links that are titled things like here "here"
```suggestion
You can find our Guide for using Jellyfin through Caddy [on the Caddy page](./caddy/).
```

Reviewer: For accessibility reasons.

Reviewer 5: ```suggestion
We recommend using [Caddy](https://caddyserver.com/) for its easy install and use. We provide a [guide](./caddy/) for configuring Caddy with Jellyfin.
```

Something like this would be even better

If you do not want to use Caddy, some other popular options for reverse proxy systems are [Nginx](https://www.nginx.com), [Traefik](https://traefik.io), [Haproxy](https://www.haproxy.com) and [Apache](https://httpd.apache.org). Note that these have a greater learning curve than the recommended reverse proxy Caddy. You can find respective guides here:

- [Nginx](./nginx)
- [Traefik](./traefik)
- [HAProxy](./haproxy)
- [Apache](./apache)

> While not a reverse proxy, Let's Encrypt can be used independently or with a reverse proxy to provide SSL certificates.

Reviewer 3: Let's encrypt can be used used in combination with a reverse proxy to provide TLS certificates.

Author: We agreed in Chat to use SSL as a synonym for TLS, as it is a common term for all http encryption and SSL standards being deprecated. Therefore marking as resolved.

- [Let's Encrypt](../advanced/letsencrypt)

When following these guides, be sure to replace the following variables with your information.

- `DOMAIN_NAME`: Your public domain name to access Jellyfin on (e.g. jellyfin.example.com)
- `example.com`: The domain name Jellyfin services will run under (e.g. example.com)
- `SERVER_IP_ADDRESS`: The IP address of your Jellyfin server (if the reverse proxy is on the same server use 127.0.0.1)

> In addition, the examples are configured for use with Let's Encrypt certificates. If you have a certificate from another source, change the SSL configuration from `/etc/letsencrypt/DOMAIN_NAME/` to the location of your certificate and key.

Reviewer 3: TLS

Reviewer 3: SSL and TLS are both cryptographic protocols used to secure communications over networks, but TLS is the successor to SSL and is considered more secure. While the terms are often used interchangeably, TLS is the current standard for secure web browsing.

Reviewer 3: ![image](https://github.com/user-attachments/assets/9c8969df-7296-4a68-a232-e9d59bcec303)

Reviewer: SSL certs are no longer used anymore. TLS certs are always used for TLS. It's just that the terms are used interchangeably
